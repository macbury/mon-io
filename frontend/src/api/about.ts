import { ApiBase } from './types'
import aboutQuery from './queries/about'
import {
  Query
} from './graphql'

export default class AboutApi extends ApiBase {
  public fetch = async () => {
    const { data } = await this.client.query<Query>({
      query: aboutQuery
    })

    return data.about
  }
}