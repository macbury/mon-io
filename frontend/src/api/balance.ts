import { ApiBase } from './types'

import {
  Query,
  QueryBalanceArgs,
  Aggregation
} from './graphql'

import getBalanceHistoryQuery from './queries/getBalanceHistory'
import getCurrentBalanceQuery from './queries/getCurrentBalance'

export default class BalanceApi extends ApiBase {
  public async getHistory(categoryId: string) {
    const { data: { balance } } = await this.client.query<Query, QueryBalanceArgs>({
      query: getBalanceHistoryQuery,
      variables: { categoryId, aggregation: Aggregation.Sum }
    })

    return balance
  }

  public async getHistoryPerMonth(categoryId: string) {
    const { data: { balance } } = await this.client.query<Query, QueryBalanceArgs>({
      query: getBalanceHistoryQuery,
      variables: { categoryId, aggregation: Aggregation.None }
    })

    return balance
  }

  public async getToday(categoryId: string) {
    const { data: { balance } } = await this.client.query<Query, QueryBalanceArgs>({
      query: getCurrentBalanceQuery,
      variables: { categoryId }
    })

    return balance?.today
  }
}
