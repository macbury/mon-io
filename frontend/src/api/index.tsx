import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createHttpLink } from "apollo-link-http"
import { BatchHttpLink } from 'apollo-link-batch-http'
import { onError, ErrorResponse } from 'apollo-link-error'
import { ApolloLink } from 'apollo-link'
import { createUploadLink } from 'apollo-upload-client'

import promiseToObservable from '../helpers/promiseToObservable'
import ApolloRefreshToken from '../helpers/ApolloRefreshToken'
import logger from '../helpers/logger'
import RootStore from '../stores/RootStore'

import BalanceApi from './balance'
import CategoriesApi from './categories'
import ReceiptsApi from './receipts'
import SessionsApi from './session'
import TransactionsApi from './transactions'
import SummaryApi from './summary'
import AboutApi from './about'
import SeriesApi from './series'
import BudgetApi from './budget'
import ImportsApi from './imports'
import LocationApi from './location'

import hasFiles from '../helpers/hasFiles'

export default class MonioApi {
  private readonly log = logger('MonioApi')

  public readonly categories : CategoriesApi
  public readonly receipts : ReceiptsApi
  public readonly sessions : SessionsApi
  public readonly transactions : TransactionsApi
  public readonly summary : SummaryApi
  public readonly about : AboutApi
  public readonly budget : BudgetApi
  public readonly series : SeriesApi
  public readonly location : LocationApi
  public readonly imports : ImportsApi
  public readonly balance : BalanceApi

  private cache : InMemoryCache
  private store : RootStore
  private apolloRefreshToken : ApolloRefreshToken

  public client : ApolloClient<any>

  constructor(store : RootStore, client? : ApolloClient<any>) {
    this.store = store

    if (client) {
      this.client = client
    } else {
      this.setupClient()
    }

    this.categories = new CategoriesApi(this.client)
    this.receipts = new ReceiptsApi(this.client)
    this.sessions = new SessionsApi(this.client)
    this.transactions = new TransactionsApi(this.client)
    this.summary = new SummaryApi(this.client)
    this.about = new AboutApi(this.client)
    this.budget = new BudgetApi(this.client)
    this.series = new SeriesApi(this.client)
    this.location = new LocationApi(this.client)
    this.imports = new ImportsApi(this.client)
    this.balance = new BalanceApi(this.client)
  }

  private get linkOptions() {
    return {
      credentials: 'omit',
      fetch: this.apiFetch,
      fetchOptions: {
        method: 'POST',
        mode: 'cors'
      }
    }
  }

  private createHttpLink() {
    if (process.env.NODE_ENV == 'development') {
      return createHttpLink(this.linkOptions)
    } else {
      return new BatchHttpLink(this.linkOptions)
    }
  }

  private setupClient() {
    this.cache = new InMemoryCache()

    const uploadLink = createUploadLink(this.linkOptions)

    const httpLink = ApolloLink.split(
      ({ variables }) => hasFiles(variables),
      uploadLink,
      this.createHttpLink(),
    );

    this.apolloRefreshToken = new ApolloRefreshToken(this.store.session)

    const link = ApolloLink.from([
      this.onLoadingLink,
      this.onErrorLink,
      httpLink
    ])

    this.client = new ApolloClient({
      link,
      cache: this.cache,
      defaultOptions: {
        query: {
          fetchPolicy: 'no-cache',
        },
        watchQuery: {
          fetchPolicy: 'no-cache',
          errorPolicy: 'all'
        }
      }
    })
  }

  private apiFetch = async (_uri, options : RequestInit) => {
    // @ts-ignore
    options.headers.Authorization = 'Token token=' + this.store.session.accessToken
    // @ts-ignore
    options.headers.MonioClientVersion = this.store.about.clientVersion
    return fetch(this.store.session.endpointUrl, options)
  }

  private onLoadingLink = new ApolloLink((operation, forward) => {
    this.store.processing = true
    return forward(operation).map(response => {
      this.store.processing = false
      return response
    })
  })

  private onErrorLink = onError(({ graphQLErrors, networkError, forward, operation } : any) => {
    if (graphQLErrors)
      graphQLErrors.forEach(({ message, locations, path }) =>
        this.log(
          `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`,
        ),
      );

    if (networkError) {
      // @ts-ignore
      const { statusCode } = networkError

      if (statusCode === 400) {
        this.log(`[Outdated api error]: ${networkError}`)
        this.store.updateRequired = true
      } else if (statusCode === 401) {
        this.log("Refreshing access token!")
        return promiseToObservable(this.apolloRefreshToken.refreshToken()).flatMap((resp) => forward(operation))
      } else {
        this.log(`[Network error]: ${networkError}`, networkError?.result)
        this.store.processing = false
        this.store.setError(networkError?.result?.error)
      }
    }
  })
}