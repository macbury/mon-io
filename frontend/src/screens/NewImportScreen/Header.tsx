import React, { useState, useCallback } from 'react'
import { StatusBar } from 'react-native'
import { useTranslation } from 'react-i18next'
import { Menu } from 'react-native-paper'
import AppHeader from '../../components/layout/AppHeader'
import MenuActionButton from '../../components/MenuActionButton'
import { useStoreData } from '../../stores'
import { transactionsPath } from '../../helpers/urls'

function useEditImport() {
  return useStoreData(({ screens: { editImport } }) => ({
    isReady: editImport.isReady,
    importId: editImport.id,
    destroy: editImport.destroy
  }))
}

export default function EditImportScreenHeader(props) {
  const [visible, setVisible] = useState(false)
  const { navigation } = props
  const { t } = useTranslation()
  const {
    isReady,
    importId,
    destroy
  } = useEditImport()

  const hideMenu = useCallback(() => setVisible(false), [setVisible])
  const showMenu = useCallback(() => setVisible(true), [setVisible])

  const destroyImport = useCallback(async () => {
    hideMenu()
    if (await destroy()) {
      navigation.pop()
    }
  }, [hideMenu, destroy, navigation])

  const showTransactions = useCallback(() => {
    hideMenu()
    navigation.dispatch(transactionsPath(null, null, importId))
  }, [hideMenu, navigation, importId])

  if (!isReady) {
    return <AppHeader title={t('pages.new_import.title')} {...props}></AppHeader>
  }

  const anchor = <MenuActionButton icon="dots-vertical" onPress={showMenu} />

  return (
    <AppHeader title={t('pages.new_import.title')} {...props}>
      <Menu onDismiss={hideMenu} visible={visible} anchor={anchor} style={{ marginTop: StatusBar.currentHeight }}>
        <Menu.Item title={t('pages.new_import.menu.show_transactions')} onPress={showTransactions} />
        <Menu.Item title={t('pages.new_import.menu.destroy')} onPress={destroyImport} />
      </Menu>
    </AppHeader>
  )
}