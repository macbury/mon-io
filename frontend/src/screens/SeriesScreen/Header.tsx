import React, { useState, useCallback } from 'react'
import { StatusBar } from 'react-native'
import { useTheme } from 'styled-components/native'
import moment from 'moment-timezone'
import { useTranslation } from 'react-i18next'
import { Menu } from 'react-native-paper'
import DateNavigationHeader from '../../components/layout/DateNavigationHeader'
import AppHeader from '../../components/layout/AppHeader'
import Mobile from '../../components/responsive/Mobile'
import Desktop from '../../components/responsive/Desktop'
import MenuActionButton from '../../components/MenuActionButton'
import { seriesPath } from '../../helpers/urls'
import { useSeriesStore } from './hooks'


export default function SeriesScreenHeader(props) {
  const { device } = useTheme()
  const { t } = useTranslation()
  const [visible, setVisible] = useState(false)
  const {
    currentDate,
    showCalendarUrl,
    reload
  } = useSeriesStore()

  const { navigation } = props

  const hideMenu = useCallback(() => setVisible(false), [setVisible])
  const showMenu = useCallback(() => setVisible(true), [setVisible])

  const shareCalendarOption = useCallback(() => {
    showCalendarUrl()
    hideMenu()
  }, [hideMenu, showCalendarUrl])

  const refreshOption = useCallback(() => {
    reload()
    hideMenu()
  }, [hideMenu, reload])

  const goToToday = useCallback(() => {
    const today = moment()
    navigation.navigate(seriesPath(today))
  }, [navigation])

  const goToDateAction = useCallback((date) => seriesPath(date), [navigation])

  const anchor = <MenuActionButton icon="dots-vertical" onPress={showMenu} />

  const Header = device === 'desktop' ? AppHeader : DateNavigationHeader

  return (
    <Header
      active={false}
      onChangeDate={goToDateAction}
      currentDate={currentDate}
      title={t('pages.series.title')}
      {...props}>
      <MenuActionButton icon="calendar" onPress={goToToday} />
      <Desktop>
        <MenuActionButton icon="refresh" onPress={refreshOption} />
      </Desktop>
      <Mobile>
        <Menu onDismiss={hideMenu} visible={visible} anchor={anchor} style={{ marginTop: StatusBar.currentHeight }}>
          <Menu.Item icon="calendar-sync" title={t('pages.series.menu.get_calendar_url')} onPress={shareCalendarOption} />
          <Menu.Item icon="refresh" title={t('pages.series.menu.refresh')} onPress={refreshOption} />
        </Menu>
      </Mobile>
    </Header>
  )
}