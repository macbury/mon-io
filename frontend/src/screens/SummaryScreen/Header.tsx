import React, { useMemo, useCallback } from 'react'
import moment, { Moment } from 'moment-timezone'
import { summaryPath } from '../../helpers/urls'
import Menu from './Menu'

import DateNavigationHeader from '../../components/layout/DateNavigationHeader'
import { useSummary } from './hooks'

export default function Header(props) {
  const { currentDate, type, } = useSummary()
  const isToday = useMemo(() => (currentDate.isSame(moment(), 'month')), [currentDate])
  const summaryNavigationAction = useCallback((date : Moment) => (summaryPath(date, type)), [type])

  return (
    <DateNavigationHeader
      active={isToday}
      currentDate={currentDate}
      onChangeDate={summaryNavigationAction}
      {...props}>
        <Menu />
    </DateNavigationHeader>
  )
}