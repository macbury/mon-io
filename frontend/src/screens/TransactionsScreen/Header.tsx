import React, { useState, useCallback } from 'react'
import styled from 'styled-components/native'
import { useTranslation } from 'react-i18next'
import { StatusBar } from 'react-native'
import { Searchbar, Menu, Text } from 'react-native-paper'
import AppHeader from '../../components/layout/AppHeader'
import MenuActionButton from '../../components/MenuActionButton'
import PortraitOrMobile from '../../components/responsive/PortraitOrMobile'
import { useStoreData } from '../../stores'

const SearchBarInput = styled(Searchbar)`
  flex: 1;
  margin: 0 auto;
  max-width: ${(props) => props.theme.device === 'desktop' ? '80%' : '100%'};
`

function useTransactions() {
  return useStoreData(({ screens: { transactions } }) => ({
    query: transactions.query,

    downloadExcel: transactions.downloadExcel,
    clearQuery: transactions.clearQuery,
    setQuery: transactions.setQuery,
    showFiltersDialog: transactions.showFiltersDialog
  }))
}

export function TransactionMenu() {
  const { t } = useTranslation()
  const {
    clearQuery,
    downloadExcel
  } = useTransactions()

  const [visible, setVisible] = useState(false)
  const anchor = <MenuActionButton icon="dots-vertical" onPress={() => setVisible(true)} />
  const hideMenu = useCallback(() => setVisible(false), [setVisible])

  const onResetFilter = useCallback(() => {
    hideMenu()
    clearQuery()
  }, [hideMenu, clearQuery])

  const onDownloadExcel = useCallback(() => {
    hideMenu()
    downloadExcel()
  }, [hideMenu, downloadExcel])

  return (
    <Menu onDismiss={hideMenu} visible={visible} anchor={anchor} style={{ marginTop: StatusBar.currentHeight }}>
      <Menu.Item title={t('pages.transactions.menu.export_excel')} onPress={onDownloadExcel} />
      <Menu.Item title={t('pages.transactions.menu.reset_filter')} onPress={onResetFilter} />
    </Menu>
  )
}

export default function TransactionsScreenHeader(props) {
  const { t } = useTranslation()
  const {
    query,
    setQuery,
    showFiltersDialog
  } = useTransactions()

  return (
    <AppHeader hideTitle={true} {...props}>
      <SearchBarInput
        placeholder={t('pages.transactions.search')}
        onChangeText={setQuery}
        value={query}
        style={{ elevation: 0 }}
      />
      <PortraitOrMobile>
        <MenuActionButton icon="filter-variant" onPress={showFiltersDialog} />
      </PortraitOrMobile>
      <TransactionMenu />
    </AppHeader>
  )
}