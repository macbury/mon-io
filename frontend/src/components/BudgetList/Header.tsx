import React, { useMemo } from 'react'
import styled from 'styled-components/native'
import { Text } from 'react-native-paper'
import { useTranslation } from 'react-i18next'
import { Money, CategoryBudget } from '../../api/graphql'
import { useFormatMoney } from '../../helpers/currency'
import { HeaderType, backgroundColorByType, textColorByType, IStyleProps } from './types'
import GoalBadge from './GoalBadge'

export interface IHeaderProps {
  type: HeaderType
  categories: CategoryBudget[]
}

const HeaderRowContainer = styled.View`
  background: ${backgroundColorByType};
  border-bottom-color: ${({ theme }) => theme.itemBorderColor};
  border-bottom-width: 1px;
  border-top-color: ${({ theme }) => theme.itemBorderColor};
  border-top-width: 1px;
`

const HeaderRowTitleAndPriceContainer = styled.View`
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 10px 20px 5px 20px;
`

const HeaderRowTitle = styled(Text)`
  font-size: 18px;
`

const ProgressBarContainer = styled.View`
  margin: 0px 20px 5px 20px;
  height: 14px;
  background: ${({ theme }) => theme.budgetProgressBackgroundColor};
  border-radius: 8px;
  border-width: 1px;
  overflow: hidden;
  border-color: ${({ theme }) => theme.budgetProgressBorderColor};
`

const ProgressValue = styled.View`
  height: 14px;
  background: ${textColorByType};
  width: ${(props : IStyleProps) => props.value * 100}%;
`

const HeaderRowSpendAndBudgeted = styled.View`
  flex-direction: row;
  justify-content: space-between;
  padding: 0px 20px 10px 20px;
`

const HeaderSpendTitle = styled(Text)`
  font-size: 14px;
  color: ${textColorByType};
`

const HeaderBudgetedTitle = styled(Text)`
  font-size: 14px;
  opacity: 0.6;
`

function sumMoney(a : Money , b : Money) : Money {
  return {
    cents: a.cents + b.cents,
    currency: a.currency
  }
}

function extractSpend(categoryBudget : CategoryBudget) {
  return categoryBudget.exchangedSpend
}

function extractPlanned(categoryBudget : CategoryBudget) {
  return categoryBudget.exchangedPlanned
}

function extractExecuted(categoryBudget : CategoryBudget) {
  return categoryBudget.executed
}

function sumNumber(a : number, b : number) {
  return a + b
}

export default function Header({ type, categories } : IHeaderProps) {
  const { t } = useTranslation()

  const spend = useMemo(() => categories.map(extractSpend).reduce(sumMoney), [categories])
  const planned = useMemo(() => categories.map(extractPlanned).reduce(sumMoney), [categories])
  const executed = useMemo(() => {
    const plannedCategoryCount = categories.filter(({ exchangedPlanned }) => (exchangedPlanned.cents > 0)).length

    if (plannedCategoryCount === 0) {
      return 0
    } else {
      return categories.map(extractExecuted).reduce(sumNumber) / plannedCategoryCount
    }
  }, [categories])

  const available = useMemo(() => ({
    cents: planned.cents - spend.cents,
    currency: planned.currency
  }), [planned, spend])

  const spendText = useFormatMoney(spend)
  const plannedText = useFormatMoney(planned)

  return (
    <HeaderRowContainer type={type}>
      <HeaderRowTitleAndPriceContainer>
        <HeaderRowTitle>{t(`pages.budget.header.${type}.title`)}</HeaderRowTitle>
        <GoalBadge
          amount={available}
          type={type} />
      </HeaderRowTitleAndPriceContainer>
      <ProgressBarContainer>
        <ProgressValue type={type} value={executed} />
      </ProgressBarContainer>
      <HeaderRowSpendAndBudgeted>
        <HeaderSpendTitle type={type}>
          {t(`pages.budget.header.${type}.current`, { amount: spendText })}
        </HeaderSpendTitle>
        <HeaderBudgetedTitle>
          {t(`pages.budget.header.${type}.budgeted`, { amount: plannedText })}
        </HeaderBudgetedTitle>
      </HeaderRowSpendAndBudgeted>
    </HeaderRowContainer>
  )
}