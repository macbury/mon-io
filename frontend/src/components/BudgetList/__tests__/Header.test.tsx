import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import { TransactionCategoryKind } from '../../../api/graphql'
import { categoryBudget } from '../../../test/factories/categoryBudget'

import Header, { IHeaderProps } from '../Header'

it('renders header with category expense for 10 zl', () => {
  const props : IHeaderProps = {
    type: TransactionCategoryKind.Expense,
    categories: categoryBudget.buildList(10)
  }

  const component = shallow(<Header {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})


it('renders header with category income for 0 zl', () => {
  const props : IHeaderProps = {
    type: TransactionCategoryKind.Income,
    categories: categoryBudget.buildList(10)
  }

  const component = shallow(<Header {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})