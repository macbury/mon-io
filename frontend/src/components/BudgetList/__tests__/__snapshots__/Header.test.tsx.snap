// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders header with category expense for 10 zl 1`] = `
<Styled(View)
  type="Expense"
>
  <Styled(View)>
    <Styled(withTheme(Text))>
      Expenses
    </Styled(withTheme(Text))>
    <GoalBadge
      amount={
        Object {
          "cents": 0,
          "currency": Object {
            "id": "PLN",
            "isoCode": "PLN",
            "name": "Polish zloty",
            "subunitToUnit": 100,
            "symbol": "zł",
          },
        }
      }
      type="Expense"
    />
  </Styled(View)>
  <Styled(View)>
    <Styled(View)
      type="Expense"
      value={3}
    />
  </Styled(View)>
  <Styled(View)>
    <Styled(withTheme(Text))
      type="Expense"
    >
      spend 10.00 zł
    </Styled(withTheme(Text))>
    <Styled(withTheme(Text))>
      budgeted 10.00 zł
    </Styled(withTheme(Text))>
  </Styled(View)>
</Styled(View)>
`;

exports[`renders header with category income for 0 zl 1`] = `
<Styled(View)
  type="Income"
>
  <Styled(View)>
    <Styled(withTheme(Text))>
      Income
    </Styled(withTheme(Text))>
    <GoalBadge
      amount={
        Object {
          "cents": 0,
          "currency": Object {
            "id": "PLN",
            "isoCode": "PLN",
            "name": "Polish zloty",
            "subunitToUnit": 100,
            "symbol": "zł",
          },
        }
      }
      type="Income"
    />
  </Styled(View)>
  <Styled(View)>
    <Styled(View)
      type="Income"
      value={3}
    />
  </Styled(View)>
  <Styled(View)>
    <Styled(withTheme(Text))
      type="Income"
    >
      received 10.00 zł
    </Styled(withTheme(Text))>
    <Styled(withTheme(Text))>
      budgeted 10.00 zł
    </Styled(withTheme(Text))>
  </Styled(View)>
</Styled(View)>
`;
